ipn/ipnlocal/profiles.go

```diff
@@ -0,0 +1,313 @@
+// Copyright (c) 2022 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package ipnlocal
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"runtime"
+	"strings"
+	"sync"
+
+	"golang.org/x/exp/slices"
+	"tailscale.com/ipn"
+	"tailscale.com/types/logger"
+)
+
+// ProfileManager is a wrapper around a StateStore that manages
+// multiple profiles and the current profile.
+type ProfileManager struct {
+	store ipn.StateStore
+	logf  logger.Logf
+
+	// Lock order: LocalBackend.mu, then pm.mu.
+	mu            sync.Mutex // guards following
+	knownProfiles []string
+	currentState  ipn.StateKey
+	prefs         ipn.PrefsView
+}
+
+// SaveAsStartState saves the provided prefs as the state to start with
+// when running in server mode. This is only used by the Windows GUI in
+// server mode.
+func (pm *ProfileManager) SaveAsStartState(userID string, prefs *ipn.Prefs) error {
+	clonedPrefs := prefs.Clone().View()
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
+	if err := pm.switchStateLocked(ipn.StateKey("user-" + userID)); err != nil {
+		return err
+	}
+	if err := pm.setPrefsLocked(clonedPrefs); err != nil {
+		return err
+	}
+	return pm.store.WriteState(ipn.ServerModeStartKey, []byte(pm.currentState))
+}
+
+// Unload unloads the current profile, if any.
+// It also clears the server-mode-start-key, so that the next time
+// tailscaled starts it will start with no profile.
+// Any future calls to SetPrefs will not be persisted, until a profile
+// is loaded.
+// This is used by the Windows in client mode, as the state is stored
+// elsewhere.
+func (pm *ProfileManager) Unload() error {
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
+	pm.prefs = emptyPrefs
+	pm.currentState = ""
+	if err := pm.useCurrentProfileAtStartup(); err != nil {
+		return err
+	}
+	return pm.store.WriteState(ipn.ServerModeStartKey, nil)
+}
+
+// SetPrefs sets the current profile's prefs to the provided value. It also
+// saves the prefs to the StateStore. If the currnet profile is named, it stores
+// a copy of the provided prefs, which maybe accessed via CurrentPrefs.
+func (pm *ProfileManager) SetPrefs(prefsIn *ipn.Prefs) error {
```

Comment:
Any reason why this doesn't take a `PrefsView` to more clearly indicate that it doesn't mutate or take ownership of the prefs?

Reply:
Good point, done.